R/epi_df.R

```diff
@@ -277,6 +277,20 @@ as_epi_df.tbl_df <- function(
   }
 
   assert_character(other_keys)
+
+  # Check one time_value per group
+  duplicated_time_values <- x %>%
+    group_by(across(all_of(c("geo_value", "time_value", other_keys)))) %>%
+    dplyr::summarize(n = dplyr::n(), .groups = "drop") %>%
+    filter(n > 1)
```

Comment:
```suggestion
  if ("n" %in% other_keys) {
    cli_abort('`other_keys` can't include "n"') 
  }
  duplicated_time_values <- x %>%
    count(across(all_of(c("geo_value", "time_value", other_keys)))) %>%
    filter(n > 1)
```
or
```suggestion
  if ("n" %in% other_keys) {
    cli_abort('`other_keys` can't include "n"') 
  }
  duplicated_time_values <- x %>%
    group_by(across(all_of(c("geo_value", "time_value", other_keys)))) %>%
    filter(dplyr::n() > 1) %>%
    ungroup()
```
if you want to actually show data

Reply:
todo: we need to error or decay to tibble when we violate this in other operations.  That means updating `dplyr_reconstruct.epi_df`.

Reply:
I'll add the guard, but I also renamed the `n` to something that will be less likely  `.time_value_counts`. Not too worried about performance rn, but we can easily refactor this section if it ever comes to that.

Reply:
> todo: we should error or decay to tibble when we violate this in other operations

we should talk through the ways this can get violated in other operations. a bit wary of decaying to tibble - silently losing epi_df-ness seems like it could lead to unpleasant surprises. will come back to this when i address other points.

Reply:
hm, getting a call stack overflow cause i use dplyr functions in the duplicated time values check above. do i really have to implement my very own group_by for this?

Reply:
still unclear how to resolve this.

Reply:
Is this still an issue?  Do you have a reprex?

Reply:
I don't, but the gist is that if you group_by an epi_df inside dplyr_reconstruct, you end up in an infinite recursion.

Reply:
I don't understand how that happens.  Aren't we working off of a tibble input & haven't converted it yet?

Reply:
Is it failing in the current state of the branch?  Or do you have some local work on `dplyr_reconstruct` that's triggering the error?

I need to re-read `dplyr_extending` and see if we actually need to touch `dplyr_reconstruct`.  We may only need it on `dplyr_row_slice` and maybe `[`.

Reply:
Caching in some stuff from memory:
- We probably don't want to do a full key validity check in dplyr_reconstruct; we run that all the time and it could be expensive.
- We might be able to get by with:
  - In `dplyr_row_slice.epi_df`: check that they have not requested the same row index twice.
  - In `[` when it selects columns: if any of the other_keys columns has been dropped, we need to do a full uniqueness check (or some optimization of it).  If it's also selecting rows... not sure if we also need that row slice check here or if row slice is already being called somehow.
  - ?dplyr_extending seems to suggest we don't need a `dplyr_col_modify` implementation for this.

Reply:
Hm, I had some local changes but I backed them out after the recursion issues. I'm going to rebase this and probably merge and we can iterate on it in the joint PR?